drivers/place/router/digraph.cr

```diff
@@ -0,0 +1,196 @@
+# Labelled digraph. Holds node labels of type *N* and edge labels of type *E*.
+#
+# Nodes are stored on UInt64 ID's. This provides an interface that should feel
+# similar to `Indexable` for interacting with nodes labels. Similarly edges can
+# be placed and retrieved by using a dual index of {predescessor, successor}.
+#
+# OPTIMIZE: replace with a sparse matrix and graphBLAS operations.
+class Place::Router::Digraph(N, E)
+  class Error < Exception; end
+
+  record Node(N, E),
+    attr : N,
+    succ : Hash(UInt64, E)
+
+  @nodes : Hash(UInt64, Node(N, E))
+
+  delegate clear, to: @nodes
+
+  def initialize(initial_capacity = nil)
+    @nodes = Hash(UInt64, Node(N, E)).new initial_capacity: initial_capacity
+  end
+
+  private def node(id)
+    node(id) { raise Error.new "Node #{id} does not exist" }
+  end
+
+  private def node(id, &)
+    id = id.to_u64
+    @nodes.fetch(id) { yield id }
+  end
+
+  private def check_node_exists(id)
+    check_node_exists(id) { raise Error.new "Node #{id} does not exist" }
+  end
+
+  private def check_node_exists(id, &)
+    id = id.to_u64
+    @nodes.has_key?(id) ? id : yield id
+  end
+
+  # Retrieves the label attached to node *id*.
+  def [](id)
+    node(id).attr
+  end
+
+  # Retrieves the label attached to node *id*. Yields if it does not exist.
+  def fetch(id, &) : N
+    node = node(id) { |id| return yield id }
+    node.attr
+  end
+
+  # Insert a new node.
+  def []=(id, attr)
+    insert(id, attr) { raise Error.new "Node #{id} already exists" }
+  end
+
+  # Inserts a node. Yields if it already exists.
+  def insert(id, attr : N, &)
+    id = id.to_u64
+    if @nodes.has_key? id
+      yield id
+    else
+      @nodes[id] = Node(N, E).new attr, {} of UInt64 => E
+    end
+  end
+
+  # Retrieves the label attached to the edge that joins *pred_id* and *succ_id*.
+  def [](pred_id, succ_id)
+    fetch(pred_id, succ_id) do
+      raise Error.new "Edge #{pred_id} -> #{succ_id} does not exist"
+    end
+  end
+
+  # :ditto:
+  def fetch(pred_id, succ_id, &) : E
+    succ_id = check_node_exists succ_id
+    node(pred_id).succ.fetch(succ_id) { yield pred_id, succ_id }
```

Comment:
Double-check that this is yielding a tuple, there was an issue where this syntax would end up yielding twice. Thought your tests probably cover this.

Reply:
Do you have an example of that oddity in action? Will triple check things here - but keen to grok why that could happen for general language understanding.

Reply:
Bump @caspiano